easybuild/tools/module_naming_scheme/toolchain.py

```diff
@@ -116,6 +116,20 @@ def det_toolchain_compilers(ec):
                 comp_elem_details['name'] = comp_elem
                 tc_comps.append(comp_elem_details)
         else:
+            if 'icc' in tc_elems[TOOLCHAIN_COMPILER] and 'ifort' in tc_elems[TOOLCHAIN_COMPILER]:
+                combined_comp_elem_details = det_toolchain_element_details(ec.toolchain, 'iccifort', allow_missing=True)
+                if combined_comp_elem_details:
+                    if isinstance(combined_comp_elem_details, EasyConfig):
+                        combined_comp_elem_details = combined_comp_elem_details.asdict()
+                    # add details for each compiler separately, using details from combo
+                    for comp_elem in ['icc', 'ifort']:
+                        comp_elem_details = copy.copy(combined_comp_elem_details)
+                        comp_elem_details['name'] = comp_elem
+                        tc_comps.append(comp_elem_details)
+                        tc_elems[TOOLCHAIN_COMPILER].remove(comp_elem)
```

Comment:
This is very copy-paste compared to what's above...

Rather than only considering the concatenation of *all* toolchain compiler elements above, we could add a special case for toolchains including both `icc` and `ifort`.

```python
tc_comp_found = False

tc_comp_elems = tc_elems[TOOLCHAIN_COMPILER]
comp_elems_list = [tc_comp_elems]

# for toolchains including icc & ifort (like iccifortcuda), always consider iccifort as well
if 'icc' in tc_comp_elems and 'ifort' in tc_comp_elems:
    comp_elems_list.append(('icc', 'ifort'))

for comp_elems in comp_elems_list:
    combined_comp_elem = ''.join(comp_elems)
    combined_comp_elem_details = ...
    if combined_comp_elem_details:
        ...
        # add details for each compiler separately, using details from combo
        for comp_elem in comp_elems:
            ...

        tc_comp_found = True
        break

if not tc_comp_found:
    # consider individual compiler module names as fallback
    ...
```

Reply:
I think doing comp_elems_list.append((''iccifort')) makes a bit more sense, but yes this is a slimmer way of doing things.

What I had in mind was to have each toolchain definition generate any concatenations to make this future proof. But then again there is currently only one case of this.
clang/flang could be a potential candidate unless that gets solved some other way.

Reply:
And the above still doesn't do what it should.
For iccifortcuda and iimpic especially you really need something more like the part I added, i.e., you have to check if 'iccifort' resolves, and then remove 'icc' and 'ifort' from the list and fall through.

So dropping the previous "first consider concatenation of list of compiler" code completely and just do something like
```
elif tc_elems[TOOLCHAIN_COMPILER]:
    tc_comps = []

    tc_comp_elems = tc_elems[TOOLCHAIN_COMPILER]

    if 'icc' in tc_comp_elems and 'ifort' in tc_comp_elems:
        combined_comp_elem_details = det_toolchain_element_details(ec.toolchain, 'iccifort', allow_missing=True)
        if combined_comp_elem_details:
            if isinstance(combined_comp_elem_details, EasyConfig):
                combined_comp_elem_details = combined_comp_elem_details.asdict()
            for comp_elem in ['icc', 'ifort']:
                comp_elem_details = copy.copy(combined_comp_elem_details)
                comp_elem_details['name'] = comp_elem
                tc_comps.append(comp_elem_details)
                tc_comp_elems..remove(comp_elem)

    # consider individual compiler module names as fallback
    for comp_elem in tc_comp_elems:
        tc_comps.append(det_toolchain_element_details(ec.toolchain, comp_elem))
```

Reply:
Rewrote it so you can look at it in context.